src/EventStore.Core/DataStructures/ObjectPool.cs

```diff
@@ -104,6 +105,7 @@ public class ObjectPool<T> : IDisposable {
 
 		public void Return(T item) {
 			_queue.Enqueue(item);
+			Thread.MemoryBarrier();
 			if (_disposing)
```

Comment:
there's something i'm not sure about:
is there any guarantee that if `_disposing` is false here, the other thread will be able to dequeue the enqueued item? or TryDequeue might also return false?

Reply:
resolving this as after discussions with @timothycoleman i realized the question wasn't completely valid.

Reply:
i tried to write something coherent (tee hee) about this below https://github.com/EventStore/EventStore/pull/3930#issuecomment-1686216888